README.md
# Body Mass Index (BMI) - Utilities
[![CI - Linter & tests](https://github.com/romenrg/body-mass-index/workflows/CI%20-%20Linter%20&%20tests/badge.svg)](https://github.com/romenrg/body-mass-index/actions?query=workflow%3A%22CI+-+Linter+%26+tests%22)
[![Coverage Status](https://coveralls.io/repos/github/romenrg/body-mass-index/badge.svg?branch=master)](https://coveralls.io/github/romenrg/body-mass-index?branch=master) 
[![PyPI version](https://badge.fury.io/py/body-mass-index.svg)](https://badge.fury.io/py/body-mass-index) 
[![GitHub license](https://img.shields.io/github/license/romenrg/body-mass-index?color=blue)](https://github.com/romenrg/body-mass-index/blob/master/LICENSE)
> Python module that provides utilities related to the Body Mass Index (BMI). Also provides a Command Line Interface (CLI). 

## Table of contents

- [Module usage](#module-usage)
    - [Install & Import](#install--import)
    - [Available functions](#available-functions)
- [Standalone usage](#standalone-usage)
    - [Run the CLI](#run-the-cli)
    - [Available CLI commands](#available-cli-commands)
- [Source code usage](#source-code-usage)
    - [Build](#build)
    - [Test](#test)
    - [Run](#run)
- [Contribute](#contribute)

## <a id="module-usage"></a>Module usage

This library can be used from Python code, imported as a Python module.

### <a id="install--import"></a>Install & Import

 * Install it with pip:
    * `pip install body-mass-index`
    * Although you might consider using `pipenv`, which creates a Pipfile to keep track of your dependencies. 
    In that case, the install command would be:
        * `pipenv install body-mass-index`
 * Then, import this module in your code:
     * `from bmi import Bmi`
     * Note that the `Bmi` class contains all the utilities of this module
 * Now, use it.
     * E.g. `Bmi.calculate_bmi(80, 1.80)`
 
### <a id="available-functions"></a>Available functions

* Two static methods:
    * `calculate_bmi`: Calculates BMI, provided weight (kg) and height (m)
    * `calculate_weight`: Calculates weight (kg), provided height (m) and BMI
* Several class methods:
    * `calculate_bmi_with_info`: Calculate BMI with range info, provided weight (kg), height (m) and language
    * `get_bmi_range_info`: Return range info, provided BMI and language
    * `get_bmi_ranges_info`: Return info of all ranges
    * `calculate_weight_boundaries`: Calculate weight boundaries (kg), based on BMI; provided height (m) 
    * `calculate_weight_ranges_with_info`: Calculate weight ranges (kg), based on BMI ranges; provided height (m) and language
    * `calculate_healthy_weight`: Calculate healthy weight range (kg), provided height (m)
* Module provides class attributes:
    * `boundaries`: List with the boundaries for BMI ranges, as defined by the WHO
    * `ranges_i18n`: Text defining each BMI range, as per the WHO, in English and Spanish
   

## <a id="standalone-usage"></a>Standalone usage

This library can also be used as a CLI utility. And it can be run both with Python or with Docker.

### <a id="run-the-cli"></a>Run the CLI

Install it with pip.

Then, you can run the CLI utility in two different ways.

#### As a Python module
`python3 -m bmi  [command] [params]`

#### As a Docker container
`docker run -it romenrg/bmi:0.0.1 [command] [params]`

Notes:
* If no command is provided, help will be displayed
* Example providing command with parameters `docker run -it romenrg/bmi:0.0.1 calculate 80 1.80`

### <a id="available-cli-commands"></a>Available CLI commands

* `calculate_bmi`: Calculate BMI, provided weight (kg) and height (m) [info]
* `get_bmi_range_info`: Return range info, provided BMI and language
* `get_bmi_ranges_info`: Return info of all BMI ranges
* `calculate_weight`: Calculate weight (kg), provided height (m) and BMI
* `calculate_weight_ranges_info`: Calculate weight ranges (kg), based on BMI ranges; provided height (m)
* `calculate_healthy_weight`: Calculate healthy weight range (kg), provided height (m)


## <a id="source-code-usage"></a>Source code usage

You can also clone the repository in order to build a Docker image, run tests, execute the module locally and/or 
contribute to keep improving this module.

### <a id="build"></a>Build

Python, as an interpreted language, doesn't have to be built, but the Docker image does.

#### Create a Docker image from sources
`docker build --tag=romenrg/bmi:0.0.1 .`

### <a id="test"></a>Test

Existing unit tests can be run in two different ways.

#### As a Python module
`python3 -m unittest`

#### As a Docker container
`docker run -it --entrypoint python3 romenrg/bmi:0.0.1 -m unittest`

### <a id="run"></a>Run

You can run the CLI from the source code in the same way you can when you installed it. See the [Run the CLI](#run-the-cli) section.


## <a id="contribute"></a>Contribute

Any contributions (i.e. PRs or issues) are welcome. Please feel free to propose changes following [the contributing guideline](https://github.com/romenrg/body-mass-index/blob/master/CONTRIBUTING.md).


changelog.md
# Changelog for Evergreen Skills for Software Developers

This file keeps a log of version releases. This file is maintained
[following best practices about changelogs](https://keepachangelog.com/en/1.0.0/).

## Versioning guideline

* Each significant change should be described in the "Unreleased" section
* Entry versions should follow the following:
  * Syntax:
    * [X.Y.Z] - YYYY-MM-DD (Short description)
  * Convention:
    * For X.Y.Z, increment the digit following this scheme:
      * X = breaking change (previous version is now deprecated),
      * Y = additive change (previous version is still valid),
      * Z = just a simple patch (for a typo or error).
* When the new version is released, a new [Unreleased] section is created in the "Release Change History" below, as a 
placeholder for the next version.

## Release Change History

### [Unreleased] [1.0.2] - YYY-MM-DD (-)

#### Added

* .

#### Changed

* .

#### Removed

* .


### [1.0.1] - 2021-01-15 (Flake8 linter & better i18n messages)

#### Added

* .

#### Changed

* Applied most Flake8 Linter suggestions (minor changes, such as breaking lines or adding spaces around operators)
* Now i18n messages for obesity use "," instead of "|" to clarify the type in each case

#### Removed

* .


### [1.0.0] - 2021-01-02 (First release, including module and CLI)

#### Added

* Module provides:
    * Two static methods:
        * `calculate_bmi`: Calculates BMI, provided weight and height
        * `calculate_weight`: Calculates weight, provided height and BMI
    * Several class methods:
        * `calculate_bmi_with_info`: Calculate BMI with range info, providing weight, height and language
        * `range_info`: Return range info, providing BMI and language
        * `ranges_with_info`: Return info of all ranges
        * `calculate_weight_ranges`: Return weight ranges, providing height 
        * `calculate_weight_ranges_with_info`: Return weight ranges info, providing height and language
        * `calculate_healthy_weight`: Return healthy weight range, providing height
    * Module provides class attributes:
        * `boundaries`: List with the boundaries for BMI ranges, as defined by the WHO
        * `ranges_i18n`: Text defining each BMI range, as per the WHO, in English and Spanish
* CLI provides several commands:
    * `calculate_bmi`: Calculate BMI, provided weight (kg) and height (m) [info]
    * `get_bmi_range_info`: Return range info, provided BMI and language
    * `get_bmi_ranges_info`: Return info of all BMI ranges
    * `calculate_weight`: Calculate weight (kg), provided height (m) and BMI
    * `calculate_weight_ranges_info`: Calculate weight ranges (kg), based on BMI ranges; provided height (m)
    * `calculate_healthy_weight`: Calculate healthy weight range (kg), provided height (m)

#### Changed

* .

#### Removed

* .


tests/test_bmi.py
import unittest
from bmi import Bmi


class TestHello(unittest.TestCase):
    def test_ranges(self):
        self.assertEqual(Bmi.boundaries, [0, 18.5, 25, 30, 35, 40])

    def test_get_bmi_range_info_under(self):
        self.assertEqual(Bmi.get_bmi_range_info(17), "Underweight")
    def test_get_bmi_range_info_healthy(self):
        self.assertEqual(Bmi.get_bmi_range_info(19), "Healthy")
    def test_get_bmi_range_info_over(self):
        self.assertEqual(Bmi.get_bmi_range_info(25.3), "Overweight")
    def test_get_bmi_range_info_obese1(self):
        self.assertEqual(Bmi.get_bmi_range_info(30.1), "Obese, Class I")
    def test_get_bmi_range_info_obese2(self):
        self.assertEqual(Bmi.get_bmi_range_info(36), "Obese, Class II")
    def test_get_bmi_range_info_obese3(self):
        self.assertEqual(Bmi.get_bmi_range_info(46), "Obese, Class III")

    def test_get_bmi_range_info_under_lang(self):
        self.assertEqual(Bmi.get_bmi_range_info(17, "es"), "Bajo peso")

    def test_get_bmi_ranges_with_info(self):
        self.assertEqual(Bmi.get_bmi_ranges_with_info(), [
            {'From': 0, 'To': 18.5, 'Info': 'Underweight'},
            {'From': 18.5, 'To': 25, 'Info': 'Healthy'},
            {'From': 25, 'To': 30, 'Info': 'Overweight'},
            {'From': 30, 'To': 35, 'Info': 'Obese, Class I'},
            {'From': 35, 'To': 40, 'Info': 'Obese, Class II'},
            {'From': 40, 'To': "", 'Info': 'Obese, Class III'},
        ])

    def test_get_bmi_ranges_with_info_lang(self):
        self.assertEqual(Bmi.get_bmi_ranges_with_info("es"), [
            {'From': 0, 'To': 18.5, 'Info': 'Bajo peso'},
            {'From': 18.5, 'To': 25, 'Info': 'Peso saludable'},
            {'From': 25, 'To': 30, 'Info': 'Sobrepeso'},
            {'From': 30, 'To': 35, 'Info': 'Obesidad, Clase I'},
            {'From': 35, 'To': 40, 'Info': 'Obesidad, Clase II'},
            {'From': 40, 'To': "", 'Info': 'Obesidad, Clase III'},
        ])

    def test_calculate_bmi_equal_80_180(self):
        self.assertEqual(Bmi.calculate_bmi(80, 1.80), 24.7, "Body Mass Index for 80kg and 1.80m is 24.7")
    def test_calculate_bmi_equal_64_164(self):
        self.assertEqual(Bmi.calculate_bmi(64, 1.64), 23.8, "Body Mass Index for 64kg and 1.64m is 23.8")
    def test_calculate_bmi_obese_93_171(self):
        self.assertEqual(Bmi.calculate_bmi(93, 1.71), 31.8, "Body Mass Index for 93kg and 1.71m is 31.8")
    def test_calculate_bmi_thin_59_176(self):
        self.assertEqual(Bmi.calculate_bmi(59, 1.76), 19, "Body Mass Index for 59kg and 1.76m is 19")

    def test_calculate_bmi_with_info(self):
        self.assertEqual(Bmi.calculate_bmi_with_info(80, 1.80), (24.7, "Healthy"))
    def test_calculate_bmi_with_info_lang(self):
        self.assertEqual(Bmi.calculate_bmi_with_info(80, 1.80, "es"), (24.7, "Peso saludable"))

    def test_calculate_weight_boundaries(self):
        self.assertEqual(Bmi.calculate_weight_boundaries(1.80), [0, 59.9, 81, 97.2, 113.4, 129.6])

    def test_calculate_weight_ranges_with_info(self):
        self.assertEqual(Bmi.calculate_weight_ranges_with_info(1.80),  [
            {'From': 0, 'To': 59.9, 'Info': 'Underweight'},
            {'From': 59.9, 'To': 81, 'Info': 'Healthy'},
            {'From': 81, 'To': 97.2, 'Info': 'Overweight'},
            {'From': 97.2, 'To': 113.4, 'Info': 'Obese, Class I'},
            {'From': 113.4, 'To': 129.6, 'Info': 'Obese, Class II'},
            {'From': 129.6, 'To': "", 'Info': 'Obese, Class III'},
        ])

    def test_calculate_weight_ranges_with_info_lang(self):
        self.assertEqual(Bmi.calculate_weight_ranges_with_info(1.80, "es"),  [
            {'From': 0, 'To': 59.9, 'Info': 'Bajo peso'},
            {'From': 59.9, 'To': 81, 'Info': 'Peso saludable'},
            {'From': 81, 'To': 97.2, 'Info': 'Sobrepeso'},
            {'From': 97.2, 'To': 113.4, 'Info': 'Obesidad, Clase I'},
            {'From': 113.4, 'To': 129.6, 'Info': 'Obesidad, Clase II'},
            {'From': 129.6, 'To': "", 'Info': 'Obesidad, Clase III'},
        ])

    def test_calculate_healthy_weight(self):
        self.assertEqual(Bmi.calculate_healthy_weight(1.80), [59.9, 81])


bmi/test/test_bmi.py
import unittest
from bmi import Bmi

class TestHello(unittest.TestCase):
    def test_ranges(self):
        self.assertEqual(Bmi.boundaries, [0, 18.5, 25, 30, 35, 40])

    def test_range_info_under(self):
        self.assertEqual(Bmi.range_info(17), "Underweight")
    def test_range_info_healthy(self):
        self.assertEqual(Bmi.range_info(19), "Healthy")
    def test_range_info_over(self):
        self.assertEqual(Bmi.range_info(25.3), "Overweight")
    def test_range_info_obese1(self):
        self.assertEqual(Bmi.range_info(30.1), "Obese | Class I")
    def test_range_info_obese2(self):
        self.assertEqual(Bmi.range_info(36), "Obese | Class II")
    def test_range_info_obese3(self):
        self.assertEqual(Bmi.range_info(46), "Obese | Class III")

    def test_range_info_under_lang(self):
        self.assertEqual(Bmi.range_info(17, "es"), "Bajo peso")

    def test_ranges_with_info(self):
        self.assertEqual(Bmi.ranges_with_info(), [
            {'From': 0, 'To': 18.5, 'Info': 'Underweight'},
            {'From': 18.5, 'To': 25, 'Info': 'Healthy'},
            {'From': 25, 'To': 30, 'Info': 'Overweight'},
            {'From': 30, 'To': 35, 'Info': 'Obese | Class I'},
            {'From': 35, 'To': 40, 'Info': 'Obese | Class II'},
            {'From': 40, 'To': "", 'Info': 'Obese | Class III'},
        ])

    def test_ranges_with_info_lang(self):
        self.assertEqual(Bmi.ranges_with_info("es"), [
            {'From': 0, 'To': 18.5, 'Info': 'Bajo peso'},
            {'From': 18.5, 'To': 25, 'Info': 'Peso saludable'},
            {'From': 25, 'To': 30, 'Info': 'Sobrepeso'},
            {'From': 30, 'To': 35, 'Info': 'Obesidad | Clase I'},
            {'From': 35, 'To': 40, 'Info': 'Obesidad | Clase II'},
            {'From': 40, 'To': "", 'Info': 'Obesidad | Clase III'},
        ])

    def test_calculate_equal_80_180(self):
        self.assertEqual(Bmi.calculate_bmi(80, 1.80), 24.7, "Body Mass Index for 80kg and 1.80m is 24.7")
    def test_calculate_equal_64_164(self):
        self.assertEqual(Bmi.calculate_bmi(64, 1.64), 23.8, "Body Mass Index for 64kg and 1.64m is 23.8")
    def test_calculate_obese_93_171(self):
        self.assertEqual(Bmi.calculate_bmi(93, 1.71), 31.8, "Body Mass Index for 93kg and 1.71m is 31.8")
    def test_calculate_thin_59_176(self):
        self.assertEqual(Bmi.calculate_bmi(59, 1.76), 19, "Body Mass Index for 59kg and 1.76m is 19")

    def test_calculate_bmi_with_info(self):
        self.assertEqual(Bmi.calculate_bmi_with_info(80, 1.80), (24.7, "Healthy"))
    def test_calculate_bmi_with_info_lang(self):
        self.assertEqual(Bmi.calculate_bmi_with_info(80, 1.80, "es"), (24.7, "Peso saludable"))

    def test_calculate_weight_ranges(self):
        self.assertEqual(Bmi.calculate_weight_boundaries(1.80), [0, 59.9, 81, 97.2, 113.4, 129.6])

    def test_calculate_weight_ranges_with_info(self):
        self.assertEqual(Bmi.calculate_weight_ranges_with_info(1.80),  [
            {'From': 0, 'To': 59.9, 'Info': 'Underweight'},
            {'From': 59.9, 'To': 81, 'Info': 'Healthy'},
            {'From': 81, 'To': 97.2, 'Info': 'Overweight'},
            {'From': 97.2, 'To': 113.4, 'Info': 'Obese | Class I'},
            {'From': 113.4, 'To': 129.6, 'Info': 'Obese | Class II'},
            {'From': 129.6, 'To': "", 'Info': 'Obese | Class III'},
        ])

    def test_calculate_weight_ranges_with_info_lang(self):
        self.assertEqual(Bmi.calculate_weight_ranges_with_info(1.80, "es"),  [
            {'From': 0, 'To': 59.9, 'Info': 'Bajo peso'},
            {'From': 59.9, 'To': 81, 'Info': 'Peso saludable'},
            {'From': 81, 'To': 97.2, 'Info': 'Sobrepeso'},
            {'From': 97.2, 'To': 113.4, 'Info': 'Obesidad | Clase I'},
            {'From': 113.4, 'To': 129.6, 'Info': 'Obesidad | Clase II'},
            {'From': 129.6, 'To': "", 'Info': 'Obesidad | Clase III'},
        ])

    def test_calculate_healthy_weight(self):
        self.assertEqual(Bmi.calculate_healthy_weight(1.80), [59.9, 81])




bmi/bmi.py
import logging.config
from os import path

logger = logging.getLogger('bmi')
logging.config.fileConfig(str(path.dirname(path.realpath(__file__))) + '/logging_config.ini')


class Bmi:

    boundaries = [
        0,
        18.5,
        25,
        30,
        35,
        40
    ]

    ranges_i18n = {
        0: {
            "en": "Underweight",
            "es": "Bajo peso"
        },
        1: {
            "en": "Healthy",
            "es": "Peso saludable"
        },
        2: {
            "en": "Overweight",
            "es": "Sobrepeso"
        },
        3: {
            "en": "Obese, Class I",
            "es": "Obesidad, Clase I"
        },
        4: {
            "en": "Obese, Class II",
            "es": "Obesidad, Clase II"
        },
        5: {
            "en": "Obese, Class III",
            "es": "Obesidad, Clase III"
        }
    }

    @staticmethod
    def calculate_bmi(weight_kg, height_m):
        exact_value = weight_kg / pow(height_m, 2)
        rounded_value = round(exact_value, 1)
        logger.debug("Calculated BMI: Exact value: " + str(exact_value) + "; Rounded value: " + str(rounded_value))
        return rounded_value

    @classmethod
    def calculate_bmi_with_info(cls, weight_kg, height_m, lang="en"):
        bmi = cls.calculate_bmi(weight_kg, height_m)
        range = cls.get_bmi_range_info(bmi, lang)
        return bmi, range

    @classmethod
    def get_bmi_range_info(cls, bmi, lang="en"):
        if bmi < cls.boundaries[1]:
            return cls.ranges_i18n[0][lang]
        elif (bmi >= cls.boundaries[1]) and (bmi < cls.boundaries[2]):
            return cls.ranges_i18n[1][lang]
        elif (bmi >= cls.boundaries[2]) and (bmi < cls.boundaries[3]):
            return cls.ranges_i18n[2][lang]
        elif (bmi >= cls.boundaries[3]) and (bmi < cls.boundaries[4]):
            return cls.ranges_i18n[3][lang]
        elif (bmi >= cls.boundaries[4]) and (bmi < cls.boundaries[5]):
            return cls.ranges_i18n[4][lang]
        else:
            return cls.ranges_i18n[5][lang]

    @classmethod
    def get_bmi_ranges_with_info(cls, lang="en"):
        detailed_ranges = []
        logger.debug("Ranges with information")
        for i, boundary in enumerate(cls.boundaries):
            if i + 1 < len(cls.boundaries):
                detailed_ranges.append({"From": boundary, "To": cls.boundaries[i + 1],
                                        "Info": cls.get_bmi_range_info(boundary, lang)})
                logger.debug("From: " + str(boundary) + "; To: " + str(cls.boundaries[i + 1]) +
                             "; Info: " + cls.get_bmi_range_info(boundary, lang))
            else:
                detailed_ranges.append({"From": boundary, "To": "", "Info": cls.get_bmi_range_info(boundary, lang)})
                logger.debug("From: " + str(boundary) + "; To: " + "; Info: " + cls.get_bmi_range_info(boundary, lang))
        return detailed_ranges

    @staticmethod
    def calculate_weight(height_m, bmi):
        weight = bmi * pow(height_m, 2)
        rounded_weight = round(weight, 1)
        logger.debug("Inputs are height: " + str(height_m) + " and bmi: " + str(bmi))
        logger.debug("Exact weight value is: " + str(weight) + "; And rounded weight: " + str(rounded_weight))
        return rounded_weight

    @classmethod
    def calculate_weight_boundaries(cls, height_m):
        weight_per_range = [
            cls.calculate_weight(height_m, cls.boundaries[0]),
            cls.calculate_weight(height_m, cls.boundaries[1]),
            cls.calculate_weight(height_m, cls.boundaries[2]),
            cls.calculate_weight(height_m, cls.boundaries[3]),
            cls.calculate_weight(height_m, cls.boundaries[4]),
            cls.calculate_weight(height_m, cls.boundaries[5])
        ]
        return weight_per_range

    @classmethod
    def calculate_weight_ranges_with_info(cls, height_m, lang="en"):
        weight_boundaries = cls.calculate_weight_boundaries(height_m)
        detailed_weight_boundaries = []
        logger.debug("Ranges weight boundaries information for height: " + str(height_m))
        for i, boundary in enumerate(weight_boundaries):
            if i + 1 < len(weight_boundaries):
                detailed_weight_boundaries.append({"From": boundary, "To": weight_boundaries[i + 1],
                                                   "Info": cls.ranges_i18n[i][lang]})
                logger.debug("From: " + str(boundary) + "; To: " + str(weight_boundaries[i + 1]) +
                             "; Info: " + cls.ranges_i18n[i][lang])
            else:
                detailed_weight_boundaries.append({"From": boundary, "To": "", "Info": cls.ranges_i18n[i][lang]})
                logger.debug("From: " + str(boundary) + "; To: " + "; Info: " + cls.ranges_i18n[i][lang])
        return detailed_weight_boundaries

    @classmethod
    def calculate_healthy_weight(cls, height_m):
        your_weight_boundaries = (cls.calculate_weight_boundaries(height_m))[1:3]
        logger.debug("For height: " + str(height_m) + ", your healthy boundaries are: " + str(your_weight_boundaries))
        return your_weight_boundaries


bmi/__main__.py
import argparse
import sys
import pprint

from bmi import Bmi


def execute_calculate_bmi(args):
    if args.info:
        return Bmi.calculate_bmi_with_info(args.weight[0], args.height[0], args.info)
    else:
        return Bmi.calculate_bmi(args.weight[0], args.height[0])


def execute_get_bmi_range_info(args):
    if args.lang:
        return Bmi.get_bmi_range_info(args.bmi[0], args.lang)
    else:
        return Bmi.get_bmi_range_info(args.bmi[0])


def execute_get_bmi_ranges_with_info(args):
    return Bmi.get_bmi_ranges_with_info(args.lang)


def execute_calculate_weight(args):
    return Bmi.calculate_weight(args.height[0], args.bmi[0])


def execute_calculate_weight_ranges_info(args):
    return Bmi.calculate_weight_ranges_with_info(args.height[0], args.lang)


def execute_calculate_healthy_weight(args):
    return Bmi.calculate_healthy_weight(args.height[0])


def subparser_is_enabled(args):
    return (len(sys.argv) > 1) and (hasattr(args, 'func'))


def cli():
    parser = argparse.ArgumentParser(description='BMI', prog='python3 -m bmi', usage='%(prog)s')
    subparsers = parser.add_subparsers(title="available commands", help='Each command provides -h option for help:')
    sub_cmd_calculate = subparsers.add_parser('calculate_bmi', help='Calculate BMI, provided weight (kg) and height '
                                                                    '(m) [info]')
    sub_cmd_calculate.set_defaults(func=execute_calculate_bmi)
    sub_cmd_calculate.add_argument('weight', metavar='weight', type=float, nargs=1,
                                   help='weight in kilograms (e.g. 80)')
    sub_cmd_calculate.add_argument('height', metavar='height', type=float, nargs=1, help='height in meters (e.g. 1.80)')
    sub_cmd_calculate.add_argument('--info', metavar='lang', type=str, nargs='?', const='en',
                                   help='Display info. Valid lang values: <"en" (default) or "es">')
    sub_cmd_range_info = subparsers.add_parser('get_bmi_range_info',
                                               help='Return range info, provided BMI and language')
    sub_cmd_range_info.set_defaults(func=execute_get_bmi_range_info)
    sub_cmd_range_info.add_argument('bmi', metavar='bmi', type=float, nargs=1,
                                    help='Body Mass Index (BMI) value (e.g. 22)')
    sub_cmd_range_info.add_argument('--lang', metavar='lang', type=str, nargs='?', default='en',
                                    help='language <"en" (default) or "es">')
    sub_cmd_ranges_w_info = subparsers.add_parser('get_bmi_ranges_info', help='Return info of all BMI ranges')
    sub_cmd_ranges_w_info.add_argument('--lang', metavar='lang', type=str, nargs='?', default='en',
                                       help='language <"en" (default) or "es">')
    sub_cmd_ranges_w_info.set_defaults(func=execute_get_bmi_ranges_with_info)
    sub_cmd_calculate_weight = subparsers.add_parser('calculate_weight',
                                                     help='Calculate weight (kg), provided height (m) and BMI')
    sub_cmd_calculate_weight.set_defaults(func=execute_calculate_weight)
    sub_cmd_calculate_weight.add_argument('height', metavar='height', type=float, nargs=1,
                                          help='height in meters (e.g. 1.80)')
    sub_cmd_calculate_weight.add_argument('bmi', metavar='bmi', type=float, nargs=1,
                                          help='Body Mass Index (BMI) value (e.g. 22)')
    sub_cmd_weight_ranges = subparsers.add_parser('calculate_weight_ranges_info',
                                                  help='Calculate weight ranges (kg), based on BMI ranges; '
                                                       'provided height (m)')
    sub_cmd_weight_ranges.set_defaults(func=execute_calculate_weight_ranges_info)
    sub_cmd_weight_ranges.add_argument('height', metavar='height', type=float, nargs=1,
                                       help='height in meters (e.g. 1.80)')
    sub_cmd_weight_ranges.add_argument('--lang', metavar='lang', type=str, nargs='?', default='en',
                                       help='language <"en" (default) or "es">')
    sub_cmd_healthy_weight = subparsers.add_parser('calculate_healthy_weight',
                                                   help='Calculate healthy weight range (kg), provided height (m)')
    sub_cmd_healthy_weight.set_defaults(func=execute_calculate_healthy_weight)
    sub_cmd_healthy_weight.add_argument('height', metavar='height', type=float, nargs=1,
                                        help='height in meters (e.g. 1.80)')
    args = parser.parse_args()
    if subparser_is_enabled(args):
        result = args.func(args)
        if isinstance(result, list):
            pp = pprint.PrettyPrinter(indent=4)
            pp.pprint(result)
        else:
            print(result)
    else:
        parser.print_help()


# Start CLI
cli()
